src/index/coinstatsindex.cpp

```diff
@@ -0,0 +1,407 @@
+        // further corruption.
+        if (m_db->Exists(DB_MUHASH)) {
+            return error("%s: Cannot read current %s state; index may be corrupted",
+                         __func__, GetName());
+        }
+
+        // If the DB_MUHASH is not set, initialize empty values
+        m_muhash = MuHash3072();
+        m_nTransactionOutputs = 0;
+        m_nBogoSize = 0;
+        m_nTotalAmount = 0;
+        m_nDiskSize = 0;
+    }
+
+    return BaseIndex::Init();
+}
+
+bool CoinStatsIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CBlockUndo block_undo;
+
+    // Ignore genesis block
+    if (pindex->nHeight > 0) {
+        if (!UndoReadFromDisk(block_undo, pindex)) {
+            return false;
+        }
+
+        std::pair<uint256, DBVal> read_out;
+        if (!m_db->Read(DBHeightKey(pindex->nHeight - 1), read_out)) {
+            return false;
+        }
+
+        uint256 expected_block_hash = pindex->pprev->GetBlockHash();
+        if (read_out.first != expected_block_hash) {
+            return error("%s: previous block header belongs to unexpected block %s; expected %s",
+                         __func__, read_out.first.ToString(), expected_block_hash.ToString());
+        }
+
+        bool is_bip30_block = (pindex->nHeight==91722 && pindex->GetBlockHash() == uint256S("0x00000000000271a2dc26e7667f8419f2e15416dc6955e5a6c6cdf3f2574dd08e")) ||
+                              (pindex->nHeight==91812 && pindex->GetBlockHash() == uint256S("0x00000000000af0aed4792b1acee3d966af36cf5def14935db8de83d6f9306f2f"));
+        MuHash3072 undo_muhash;
+
+        // Add the new utxos created from the block
+        for (size_t i = 0; i < block.vtx.size(); ++i) {
+            const auto& tx = block.vtx.at(i);
+
+            // Skip duplicate txid coinbase transactions (BIP30).
+            if (is_bip30_block && tx->IsCoinBase()) {
+                continue;
+            }
+
+            for (size_t j = 0; j < tx->vout.size(); ++j) {
+                const CTxOut& out = tx->vout[j];
+                COutPoint outpoint = COutPoint(tx->GetHash(), j);
+                Coin coin = Coin(out, pindex->nHeight, tx->IsCoinBase());
+
+                // Skip unspendable coins
+                if (coin.out.scriptPubKey.IsUnspendable()) continue;
+
+                TruncatedSHA512Writer ss(SER_DISK, 0);
+                ss << outpoint;
+                ss << (uint32_t)(coin.nHeight * 2 + coin.fCoinBase);
+                ss << coin.out;
+                m_muhash *= MuHash3072(ss.GetHash().begin());
+
+                m_nTransactionOutputs++;
+                m_nTotalAmount += coin.out.nValue;
+                m_nBogoSize += 32 /* txid */ + 4 /* vout index */ + 4 /* height + coinbase */ + 8 /* amount */ +
+                               2 /* scriptPubKey len */ + coin.out.scriptPubKey.size() /* scriptPubKey */;
+            }
+
+            // The coinbase tx has no undo data since no former output is spent
```

Comment:
I'm confused: the coinbase _is_ added to the muhash accumulator.

Reply:
Yes but the coinbase does not have any undo data because it is not spending any outputs. This section is only dealing with the undo data of the transactions in the block, it is not dealing with reorgs etc. which are further below.